{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "27ec0e08-7b27-47eb-94db-de745054e484",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gtfparse import read_gtf\n",
    "import pandas as pd\n",
    "from collections import defaultdict, OrderedDict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "511f402f-b8c0-45db-a744-129f2792a8cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bed_intersect_gen25-merged\n",
      "Combing_GTFs\n",
      "gencode.vM18.primary_assembly.annotation.gtf\n",
      "gencode.vM25.primary_assembly.annotation.gff3\n",
      "gencode.vM25.primary_assembly.annotation.gff3.gz\n",
      "gencode.vM25.primary.gtf\n",
      "GSE141750_Mse_nanopore.isoforms.gtf\n",
      "GSE141750_Mse_nanopore.isoforms.gtf.gz\n",
      "isoform_productivity.gtf\n",
      "isoform_productivity.gz\n",
      "merged_comp.gtf\n",
      "Untitled.ipynb\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "391697ab-c581-4ec2-b36d-e31fcd348036",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrs = [\"chr\" + str(num) for num in range(1,21)] + ['chrY', 'chrM', 'chrX']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "c0311243-751b-47fa-93f0-83580143572b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'protein_id', 'ccdsid', 'ont']\n"
     ]
    }
   ],
   "source": [
    "#gencode = read_gtf(\"gencode.vM25.primary.gtf\")\n",
    "gencode = read_gtf(\"gencode.vM18.primary_assembly.annotation.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "bc8dcf0a-88e1-48d1-a5dc-596d6d602943",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode.drop(['source', 'frame', 'gene_id', 'gene_type', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag',\n",
    "             'havana_transcript', 'exon_number', 'exon_id', 'protein_id', 'ccdsid', 'ont', 'level', 'havana_gene', \n",
    "              'transcript_id', 'score'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "50692ec5-071d-44d6-a7dc-5d7bc0873343",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode = gencode[gencode['feature'].isin(['exon'])]\n",
    "gencode = gencode[gencode['seqname'].isin(chrs)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "4a8a1617-a1c3-46c7-91f9-ad6b22739181",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>+</td>\n",
       "      <td>RP23-271O17.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>+</td>\n",
       "      <td>Gm26206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213609</td>\n",
       "      <td>3216344</td>\n",
       "      <td>-</td>\n",
       "      <td>Xkr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3207317</td>\n",
       "      <td>-</td>\n",
       "      <td>Xkr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213439</td>\n",
       "      <td>3215632</td>\n",
       "      <td>-</td>\n",
       "      <td>Xkr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1796856</th>\n",
       "      <td>chrM</td>\n",
       "      <td>exon</td>\n",
       "      <td>13552</td>\n",
       "      <td>14070</td>\n",
       "      <td>-</td>\n",
       "      <td>mt-Nd6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1796863</th>\n",
       "      <td>chrM</td>\n",
       "      <td>exon</td>\n",
       "      <td>14071</td>\n",
       "      <td>14139</td>\n",
       "      <td>-</td>\n",
       "      <td>mt-Te</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1796866</th>\n",
       "      <td>chrM</td>\n",
       "      <td>exon</td>\n",
       "      <td>14145</td>\n",
       "      <td>15288</td>\n",
       "      <td>+</td>\n",
       "      <td>mt-Cytb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1796871</th>\n",
       "      <td>chrM</td>\n",
       "      <td>exon</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>+</td>\n",
       "      <td>mt-Tt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1796874</th>\n",
       "      <td>chrM</td>\n",
       "      <td>exon</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>-</td>\n",
       "      <td>mt-Tp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>807591 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname feature    start      end strand      gene_name\n",
       "2          chr1    exon  3073253  3074322      +  RP23-271O17.1\n",
       "5          chr1    exon  3102016  3102125      +        Gm26206\n",
       "8          chr1    exon  3213609  3216344      -           Xkr4\n",
       "9          chr1    exon  3205901  3207317      -           Xkr4\n",
       "11         chr1    exon  3213439  3215632      -           Xkr4\n",
       "...         ...     ...      ...      ...    ...            ...\n",
       "1796856    chrM    exon    13552    14070      -         mt-Nd6\n",
       "1796863    chrM    exon    14071    14139      -          mt-Te\n",
       "1796866    chrM    exon    14145    15288      +        mt-Cytb\n",
       "1796871    chrM    exon    15289    15355      +          mt-Tt\n",
       "1796874    chrM    exon    15356    15422      -          mt-Tp\n",
       "\n",
       "[807591 rows x 6 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "459a8abc-b08f-437a-867f-e07121d5f609",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'exon_number', 'oId', 'tss_id']\n"
     ]
    }
   ],
   "source": [
    "merged_comp = read_gtf(\"merged_comp.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "26459871-84ab-438c-860d-36386f832152",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_comp.drop(['source', 'frame', 'exon_number', 'oId', 'tss_id', 'score'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "5495dda7-2704-4592-92f3-86fb84e603a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>+</td>\n",
       "      <td>XLOC_000001</td>\n",
       "      <td>TCONS_00000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>+</td>\n",
       "      <td>XLOC_000002</td>\n",
       "      <td>TCONS_00000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3680155</td>\n",
       "      <td>3681788</td>\n",
       "      <td>+</td>\n",
       "      <td>XLOC_000003</td>\n",
       "      <td>TCONS_00000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3752010</td>\n",
       "      <td>3754360</td>\n",
       "      <td>+</td>\n",
       "      <td>XLOC_000004</td>\n",
       "      <td>TCONS_00000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>4496551</td>\n",
       "      <td>4499378</td>\n",
       "      <td>+</td>\n",
       "      <td>XLOC_000005</td>\n",
       "      <td>TCONS_00000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999707</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>90665346</td>\n",
       "      <td>90665414</td>\n",
       "      <td>-</td>\n",
       "      <td>XLOC_048379</td>\n",
       "      <td>TCONS_00135677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999708</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>90666856</td>\n",
       "      <td>90667037</td>\n",
       "      <td>-</td>\n",
       "      <td>XLOC_048379</td>\n",
       "      <td>TCONS_00135677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999709</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>90667525</td>\n",
       "      <td>90667625</td>\n",
       "      <td>-</td>\n",
       "      <td>XLOC_048379</td>\n",
       "      <td>TCONS_00135677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999710</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>90754513</td>\n",
       "      <td>90754821</td>\n",
       "      <td>-</td>\n",
       "      <td>XLOC_048380</td>\n",
       "      <td>TCONS_00135679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999711</th>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>90838869</td>\n",
       "      <td>90839177</td>\n",
       "      <td>-</td>\n",
       "      <td>XLOC_048381</td>\n",
       "      <td>TCONS_00135680</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>999712 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname feature     start       end strand      gene_id   transcript_id\n",
       "0         chr1    exon   3073253   3074322      +  XLOC_000001  TCONS_00000001\n",
       "1         chr1    exon   3102016   3102125      +  XLOC_000002  TCONS_00000002\n",
       "2         chr1    exon   3680155   3681788      +  XLOC_000003  TCONS_00000003\n",
       "3         chr1    exon   3752010   3754360      +  XLOC_000004  TCONS_00000004\n",
       "4         chr1    exon   4496551   4499378      +  XLOC_000005  TCONS_00000005\n",
       "...        ...     ...       ...       ...    ...          ...             ...\n",
       "999707    chrY    exon  90665346  90665414      -  XLOC_048379  TCONS_00135677\n",
       "999708    chrY    exon  90666856  90667037      -  XLOC_048379  TCONS_00135677\n",
       "999709    chrY    exon  90667525  90667625      -  XLOC_048379  TCONS_00135677\n",
       "999710    chrY    exon  90754513  90754821      -  XLOC_048380  TCONS_00135679\n",
       "999711    chrY    exon  90838869  90839177      -  XLOC_048381  TCONS_00135680\n",
       "\n",
       "[999712 rows x 7 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_comp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "b60122ee-ad41-4bc1-b60a-e7953c9de3f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_comp.to_csv(\"XLOC.tsv\",index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "022e8c13-fe6a-4a7e-b02f-c76ff4b46fb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = pd.merge(gencode, merged_comp,  how='outer', left_on=['start','end'], right_on = ['start','end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "1e223e55-6933-404f-a1bd-c790806bc714",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>feature_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand_x</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>feature_y</th>\n",
       "      <th>strand_y</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>+</td>\n",
       "      <td>RP23-271O17.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>+</td>\n",
       "      <td>XLOC_000001</td>\n",
       "      <td>TCONS_00000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>+</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>+</td>\n",
       "      <td>XLOC_000002</td>\n",
       "      <td>TCONS_00000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213609</td>\n",
       "      <td>3216344</td>\n",
       "      <td>-</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3207317</td>\n",
       "      <td>-</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>-</td>\n",
       "      <td>XLOC_001549</td>\n",
       "      <td>TCONS_00004177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3207317</td>\n",
       "      <td>-</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>-</td>\n",
       "      <td>XLOC_001549</td>\n",
       "      <td>TCONS_00004176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2855966</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>73346504</td>\n",
       "      <td>73347038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>-</td>\n",
       "      <td>XLOC_048158</td>\n",
       "      <td>TCONS_00135437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2855967</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>73346504</td>\n",
       "      <td>73347038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>-</td>\n",
       "      <td>XLOC_048158</td>\n",
       "      <td>TCONS_00135443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2855968</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74041330</td>\n",
       "      <td>74041861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>-</td>\n",
       "      <td>XLOC_048166</td>\n",
       "      <td>TCONS_00135455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2855969</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74041330</td>\n",
       "      <td>74041861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>-</td>\n",
       "      <td>XLOC_048166</td>\n",
       "      <td>TCONS_00135454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2855970</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>85528517</td>\n",
       "      <td>85529624</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY</td>\n",
       "      <td>exon</td>\n",
       "      <td>-</td>\n",
       "      <td>XLOC_048320</td>\n",
       "      <td>TCONS_00135615</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2855971 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname_x feature_x     start       end strand_x      gene_name  \\\n",
       "0            chr1      exon   3073253   3074322        +  RP23-271O17.1   \n",
       "1            chr1      exon   3102016   3102125        +        Gm26206   \n",
       "2            chr1      exon   3213609   3216344        -           Xkr4   \n",
       "3            chr1      exon   3205901   3207317        -           Xkr4   \n",
       "4            chr1      exon   3205901   3207317        -           Xkr4   \n",
       "...           ...       ...       ...       ...      ...            ...   \n",
       "2855966       NaN       NaN  73346504  73347038      NaN            NaN   \n",
       "2855967       NaN       NaN  73346504  73347038      NaN            NaN   \n",
       "2855968       NaN       NaN  74041330  74041861      NaN            NaN   \n",
       "2855969       NaN       NaN  74041330  74041861      NaN            NaN   \n",
       "2855970       NaN       NaN  85528517  85529624      NaN            NaN   \n",
       "\n",
       "        seqname_y feature_y strand_y      gene_id   transcript_id  \n",
       "0            chr1      exon        +  XLOC_000001  TCONS_00000001  \n",
       "1            chr1      exon        +  XLOC_000002  TCONS_00000002  \n",
       "2             NaN       NaN      NaN          NaN             NaN  \n",
       "3            chr1      exon        -  XLOC_001549  TCONS_00004177  \n",
       "4            chr1      exon        -  XLOC_001549  TCONS_00004176  \n",
       "...           ...       ...      ...          ...             ...  \n",
       "2855966      chrY      exon        -  XLOC_048158  TCONS_00135437  \n",
       "2855967      chrY      exon        -  XLOC_048158  TCONS_00135443  \n",
       "2855968      chrY      exon        -  XLOC_048166  TCONS_00135455  \n",
       "2855969      chrY      exon        -  XLOC_048166  TCONS_00135454  \n",
       "2855970      chrY      exon        -  XLOC_048320  TCONS_00135615  \n",
       "\n",
       "[2855971 rows x 11 columns]"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "5cd9dce3-c566-4ff6-a773-bffe72ec269b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>feature_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand_x</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>feature_y</th>\n",
       "      <th>strand_y</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2590794</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37485851</td>\n",
       "      <td>37488982</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>-</td>\n",
       "      <td>XLOC_001812</td>\n",
       "      <td>TCONS_00004789</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname_x feature_x     start       end strand_x gene_name seqname_y  \\\n",
       "2590794       NaN       NaN  37485851  37488982      NaN       NaN      chr1   \n",
       "\n",
       "        feature_y strand_y      gene_id   transcript_id  \n",
       "2590794      exon        -  XLOC_001812  TCONS_00004789  "
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged[merged['gene_id'] == \"XLOC_001812\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "id": "209ab582-1b44-475a-84f1-cdf3fd32fd1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>feature_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand_x</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>feature_y</th>\n",
       "      <th>strand_y</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2513716</th>\n",
       "      <td>chrX</td>\n",
       "      <td>exon</td>\n",
       "      <td>60706270</td>\n",
       "      <td>60707063</td>\n",
       "      <td>-</td>\n",
       "      <td>RP23-6C3.2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>exon</td>\n",
       "      <td>nan</td>\n",
       "      <td>XLOC_046825</td>\n",
       "      <td>TCONS_00133540</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname_x feature_x     start       end strand_x   gene_name  \\\n",
       "2513716      chrX      exon  60706270  60707063        -  RP23-6C3.2   \n",
       "\n",
       "        seqname_y feature_y strand_y      gene_id   transcript_id  \n",
       "2513716      chrX      exon      nan  XLOC_046825  TCONS_00133540  "
      ]
     },
     "execution_count": 243,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged[merged['gene_name'] == \"RP23-6C3.2\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "6eaa3b70-51a7-4220-b39c-8b8f47d50ddc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#merged[~merged['gene_name'].notna()]\n",
    "#merged[merged['transcript_id'] == 'TCONS_00000011']\n",
    "reduced = merged.drop(['seqname_x', 'feature_x', 'start', 'end', 'strand_x', 'strand_y', 'seqname_y', 'feature_y', 'transcript_id'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "73590edf-1ba3-462f-8cf6-0bb316e4b8d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "54088"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#len(reduced.drop_duplicates()['gene_name'].tolist())\n",
    "len(set(reduced.drop_duplicates()['gene_name'].tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "id": "314e1103-9f10-484a-a086-255ae012283a",
   "metadata": {},
   "outputs": [],
   "source": [
    "reduced.drop_duplicates(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "42c28ab2-e627-4c62-966f-97521ad6dcbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "xlocs = reduced.drop_duplicates()['gene_id'].tolist()\n",
    "xlocs = sorted([i for i in xlocs if not pd.isna(i)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "dcf57592-9721-4032-827b-ac42d54d5e5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "translation_dict = defaultdict(list)\n",
    "\n",
    "for i in xlocs:\n",
    "    id_list = (reduced[reduced['gene_id'] == i]['gene_name'].tolist())#(index=False))\n",
    "    for l in id_list:\n",
    "        if l not in translation_dict[i] and not pd.isna(l):\n",
    "            translation_dict[i].append(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "id": "e67ef450-03f5-4486-ab12-3ca1bf164295",
   "metadata": {},
   "outputs": [],
   "source": [
    "sort_t_dict = OrderedDict(sorted(translation_dict.items()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "68efbfb3-9330-4ba6-bacd-e111929d6dc2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 253,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(xlocs)) == len(sort_t_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "0dc4e0a0-77b3-4e04-aa40-c19c9e21e177",
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0\n",
    "with open(\"XLOC_to_Gencode_18.txt\", 'w') as f:\n",
    "    for k,v in sort_t_dict.items():\n",
    "        f.write(k + \"\\t\" + \",\".join(v) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "916d96bf-a2bc-4511-bfcb-4c349fdaf202",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_matched(tsv_in):\n",
    "    with open (tsv_in, 'r') as f:\n",
    "        matched_count, missing_count = 0, 0\n",
    "        for s in f.readlines():\n",
    "            s = s.strip().split(\"\\t\")\n",
    "            if len(s) == 1: missing_count+=1\n",
    "            else: matched_count+=1\n",
    "        print(\"Missing:\\t{}\\nMatched:\\t{}\\nPercent Matched\\t{:.2f}%\".format(missing_count, matched_count, (matched_count - missing_count)/matched_count * 100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "171e17ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing:\t1245\n",
      "Matched:\t47136\n",
      "Percent Matched\t97.36%\n"
     ]
    }
   ],
   "source": [
    "count_matched('XLOC_to_Gencode_18.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8e4c2ad0",
   "metadata": {},
   "outputs": [],
   "source": [
    "convert_to_sed('XLOC_to_Gencode_18.txt', 'XLOC_to_Gencode_18_SED')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "17d28d3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_to_sed(tsv_in, sed_out):\n",
    "    str_out=\"s/{}/{}/g\\n\"\n",
    "    with open (sed_out, 'w+') as o:\n",
    "        with open (tsv_in, 'r') as f:\n",
    "            entries = [i.strip().split(\"\\t\") for i in f.readlines()]\n",
    "            o.writelines([str_out.format(e[0], e[0]) if len(e) == 1 else str_out.format(e[0], e[1]) for e in entries])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "8c806a88",
   "metadata": {},
   "outputs": [],
   "source": [
    "convert_to_sed('XLOC_to_Gencode_18.txt', 'XLOC_to_Gencode_18_SED')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
